File: rust/wxdragon-sys/cpp/src/imagelist.cpp
#include "../include/wxdragon.h"
#include <wx/imaglist.h>
#include <wx/bitmap.h>

// Helper to cast to wxImageList*
static inline wxImageList* ToWxImageList(wxd_ImageList_t* self) {
    return reinterpret_cast<wxImageList*>(self);
}

// Helper to cast to wxBitmap*
static inline wxBitmap* ToWxBitmap(wxd_Bitmap_t* self) {
    return reinterpret_cast<wxBitmap*>(self);
}

WXD_EXPORTED wxd_ImageList_t* wxd_ImageList_Create(int width, int height, bool mask, int initialCount) {
    if (width <= 0 || height <= 0) {
        // wxImageList asserts if width/height is 0, better to return null early.
        return nullptr;
    }
    wxImageList* list = new wxImageList(width, height, mask, initialCount);
    return reinterpret_cast<wxd_ImageList_t*>(list);
}

WXD_EXPORTED void wxd_ImageList_Destroy(wxd_ImageList_t* self) {
    delete ToWxImageList(self);
}

WXD_EXPORTED int wxd_ImageList_Add(wxd_ImageList_t* self, wxd_Bitmap_t* bitmap_ptr) {
    wxImageList* list = ToWxImageList(self);
    wxBitmap* bitmap = ToWxBitmap(bitmap_ptr);
    if (!list || !bitmap || !bitmap->IsOk()) {
        return -1;
    }
    // wxImageList::Add copies the bitmap
    return list->Add(*bitmap);
}

WXD_EXPORTED int wxd_ImageList_AddWithMask(wxd_ImageList_t* self, wxd_Bitmap_t* bitmap_ptr, wxd_Bitmap_t* mask_ptr) {
    wxImageList* list = ToWxImageList(self);
    wxBitmap* bitmap = ToWxBitmap(bitmap_ptr);
    wxBitmap* mask = ToWxBitmap(mask_ptr);

    if (!list || !bitmap || !bitmap->IsOk()) {
        return -1;
    }
    if (!mask || !mask->IsOk()) { // Mask must also be valid
        return -1;
    }
    // wxImageList::Add copies both bitmaps
    return list->Add(*bitmap, *mask);
}

WXD_EXPORTED int wxd_ImageList_GetImageCount(wxd_ImageList_t* self) {
    wxImageList* list = ToWxImageList(self);
    if (!list) {
        return 0; // Or -1 to indicate error, but 0 is typical for empty/null
    }
    return list->GetImageCount();
}

WXD_EXPORTED bool wxd_ImageList_RemoveAll(wxd_ImageList_t* self) {
    wxImageList* list = ToWxImageList(self);
    if (!list) {
        return false;
    }
    return list->RemoveAll();
} 

File: rust/wxdragon/src/widgets/imagelist.rs
//! ImageList widget wrapper.

use wxdragon_sys as ffi;
use crate::bitmap::Bitmap; // Corrected path: crate::bitmap instead of crate::widgets::bitmap

/// Represents a wxImageList.
///
/// This struct is a wrapper around the wxWidgets ImageList object.
/// For now, it primarily serves to provide a typed wrapper for the raw pointer.
///
/// TODO:
/// - Implement a `new()` constructor that calls an FFI function like `wxd_ImageList_new()`.
/// - Implement a `Drop` trait to call `wxd_ImageList_Destroy()` for owned instances.
/// - Implement methods to add images (e.g., `add(&self, bitmap: &Bitmap) -> i32`).
#[derive(Debug)]
pub struct ImageList {
    pub(crate) ptr: *mut ffi::wxd_ImageList_t,
    /// Indicates whether Rust owns this ImageList and should destroy it on Drop.
    /// `false` if this ImageList is wrapping a pointer obtained from wxWidgets
    /// that wxWidgets still owns (e.g., via `get_image_list`).
    pub(crate) owned: bool,
}

impl ImageList {
    /// Creates a new, empty image list.
    ///
    /// # Arguments
    /// * `width` - The width of images in the list.
    /// * `height` - The height of images in the list.
    /// * `mask` - `true` to use a mask, `false` otherwise.
    /// * `initial_count` - The initial number of images the list can store (can be resized).
    ///
    /// Returns `Some(ImageList)` on success, or `None` if creation failed (e.g. invalid dimensions).
    pub fn new(width: i32, height: i32, mask: bool, initial_count: i32) -> Option<Self> {
        if width <= 0 || height <= 0 {
            return None; // Match C++ side check
        }
        let ptr = unsafe { ffi::wxd_ImageList_Create(width, height, mask, initial_count) };
        if ptr.is_null() {
            None
        } else {
            Some(ImageList { ptr, owned: true })
        }
    }

    /// Creates a new `ImageList` from a raw pointer, for an ImageList whose lifetime is managed by wxWidgets.
    /// Rust will not attempt to destroy this ImageList on Drop.
    /// This is typically used internally when retrieving an ImageList from a control.
    pub(crate) unsafe fn from_ptr_unowned(ptr: *mut ffi::wxd_ImageList_t) -> Self {
        ImageList { ptr, owned: false }
    }

    /// Returns the raw pointer to the underlying wxImageList object.
    ///
    /// This pointer can be used for FFI calls. The caller must ensure the
    /// `ImageList` object (and thus the pointer) remains valid.
    pub fn as_ptr(&self) -> *mut ffi::wxd_ImageList_t {
        self.ptr
    }

    /// Adds a bitmap to the image list.
    /// The `wxImageList` makes an internal copy of the bitmap.
    ///
    /// # Arguments
    /// * `bitmap` - The bitmap to add.
    ///
    /// Returns the index of the added image, or -1 if an error occurred.
    pub fn add_bitmap(&self, bitmap: &Bitmap) -> i32 {
        if self.ptr.is_null() || bitmap.as_ptr().is_null() {
            return -1;
        }
        unsafe { ffi::wxd_ImageList_Add(self.ptr, bitmap.as_ptr()) }
    }

    /// Adds a bitmap and its mask to the image list.
    /// The `wxImageList` makes internal copies of both the bitmap and the mask.
    ///
    /// # Arguments
    /// * `bitmap` - The bitmap to add.
    /// * `mask` - The mask for the bitmap.
    ///
    /// Returns the index of the added image, or -1 if an error occurred.
    pub fn add_bitmap_with_mask(&self, bitmap: &Bitmap, mask: &Bitmap) -> i32 {
        if self.ptr.is_null() || bitmap.as_ptr().is_null() || mask.as_ptr().is_null() {
            return -1;
        }
        unsafe { ffi::wxd_ImageList_AddWithMask(self.ptr, bitmap.as_ptr(), mask.as_ptr()) }
    }

    /// Returns the number of images in the list.
    pub fn get_image_count(&self) -> i32 {
        if self.ptr.is_null() {
            return 0; // Consistent with C++ returning 0 for null list
        }
        unsafe { ffi::wxd_ImageList_GetImageCount(self.ptr) }
    }

    /// Removes all images from the list.
    ///
    /// Returns `true` if successful, `false` otherwise (e.g., if the list pointer is null).
    pub fn remove_all(&self) -> bool {
        if self.ptr.is_null() {
            return false;
        }
        unsafe { ffi::wxd_ImageList_RemoveAll(self.ptr) }
    }
}

impl Drop for ImageList {
    fn drop(&mut self) {
        if self.owned && !self.ptr.is_null() {
            unsafe { ffi::wxd_ImageList_Destroy(self.ptr); }
        }
    }
}

// Ensure ImageList is Send + Sync if it makes sense.
// wxWidgets objects are generally not Send/Sync unless explicitly designed for it.
// Since this wraps a raw pointer tied to the main UI thread, it's likely !Send and !Sync.
// For now, no explicit impl/negative impl. Default is !Send/!Sync due to raw pointer. 